File: infra-ts/components/registry.ts
import * as pulumi from "@pulumi/pulumi";
import * as gcp from "@pulumi/gcp";

export class ArtifactRegistry extends pulumi.ComponentResource {
  public readonly registryUrl: pulumi.Output<string>;

  constructor(
    name: string,
    args: { projectId: string; region: string },
    opts?: pulumi.ComponentResourceOptions,
  ) {
    super("genomics:infra:ArtifactRegistry", name, {}, opts);

    const repo = new gcp.artifactregistry.Repository(`${name}-repo`, {
      repositoryId: "precision-genomics",
      project: args.projectId,
      location: args.region,
      format: "DOCKER",
      labels: { service: "precision-genomics" },
    }, { parent: this });

    this.registryUrl = pulumi.interpolate`${args.region}-docker.pkg.dev/${args.projectId}/${repo.repositoryId}`;

    this.registerOutputs({ registryUrl: this.registryUrl });
  }
}

File: infra/components/registry.py
"""Artifact Registry for Docker images."""

import pulumi
import pulumi_gcp as gcp


class ArtifactRegistry(pulumi.ComponentResource):
    """Docker container registry in Artifact Registry."""

    registry_url: pulumi.Output[str]

    def __init__(
        self,
        name: str,
        project_id: str,
        region: str,
        opts: pulumi.ResourceOptions | None = None,
    ):
        super().__init__("genomics:infra:ArtifactRegistry", name, None, opts)

        self.repo = gcp.artifactregistry.Repository(
            f"{name}-repo",
            project=project_id,
            location=region,
            repository_id="precision-genomics",
            format="DOCKER",
            description="Docker images for Precision Genomics Agent Platform",
            opts=pulumi.ResourceOptions(parent=self),
        )

        self.registry_url = pulumi.Output.concat(
            region, "-docker.pkg.dev/", project_id, "/", self.repo.repository_id
        )

        self.register_outputs({"registry_url": self.registry_url})
